src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1455,6 +1453,39 @@ public ValidateSetAttribute(params string[] validValues)
 
             _validValues = validValues;
         }
+
+        /// <summary>
+        /// Initializes a new instance of the ValidateSetAttribute class
+        /// </summary>
+        /// <param name="valuesGeneratorType">type that implements the 'IValidateSetValuesGenerator' interface</param>
+        /// <exception cref="ArgumentNullException">for null arguments</exception>
+        /// <exception cref="ArgumentOutOfRangeException">for invalid arguments</exception>
+        public ValidateSetAttribute(Type valuesGeneratorType)
+        {
+            IValidateSetValuesGenerator validValuesGenerator;
+
+            if (typeof(IValidateSetValuesGenerator).IsAssignableFrom(valuesGeneratorType))
+            {
+                validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(valuesGeneratorType);
+            }
+            else
+            {
+                throw PSTraceSource.NewArgumentException("valuesGeneratorType");
+            }
+
+            _validValues = validValuesGenerator.GetValidValues();
```

Comment:
The constructor should just save `valuesGeneratorType` because we create just one instance of the attribute per use.

For example, if someone added `ValidateSet` to the `Name` parameter on `Get-VM` - you would want the name of each vm whenever you called `Get-VM`, not the names when you first called `Get-VM`.

By constructing the instance and getting the values during validation, the set is dynamically generated when tab completion or validation is happening - which only aligns with construction once (for scripts, attributes are constructed at first invocation) or never (for binary cmdlets, attributes are eagerly constructed).